wrappers/python/simtk/openmm/app/forcefield.py

```diff
@@ -1875,24 +1877,59 @@ def createForce(self, sys, data, nonbondedMethod, nonbondedCutoff, args):
                 types3 = tordef.types3
                 types4 = tordef.types4
                 hasWildcard = (wildcard in (types1, types2, types3, types4))
+                if (types1, types2, types3, types4).count(wildcard) == 2:
+                    hasTwoWildcards = True
+                else:
+                    hasTwoWildcards = False
                 if match is not None and hasWildcard:
                     # Prefer specific definitions over ones with wildcards
                     continue
                 if type1 in types1:
                     for (t2, t3, t4) in itertools.permutations(((type2, 1), (type3, 2), (type4, 3))):
                         if t2[0] in types2 and t3[0] in types3 and t4[0] in types4:
-                            # Workaround to be more consistent with AMBER.  It uses wildcards to define most of its
-                            # impropers, which leaves the ordering ambiguous.  It then follows some bizarre rules
-                            # to pick the order.
+                            # topology atom indexes
                             a1 = torsion[t2[1]]
                             a2 = torsion[t3[1]]
+                            a4 = torsion[t4[1]]
+                            # residue indexes
+                            r1 = data.atoms[a1].residue.index
+                            r2 = data.atoms[a2].residue.index
+                            r4 = data.atoms[a4].residue.index
+                            # template atom indexes
+                            ta1 = data.atomTemplateIndexes[data.atoms[a1]]
+                            ta2 = data.atomTemplateIndexes[data.atoms[a2]]
+                            ta4 = data.atomTemplateIndexes[data.atoms[a4]]
+                            # elements
                             e1 = data.atoms[a1].element
                             e2 = data.atoms[a2].element
-                            if e1 == e2 and a1 > a2:
-                                (a1, a2) = (a2, a1)
-                            elif e1 != elem.carbon and (e2 == elem.carbon or e1.mass < e2.mass):
-                                (a1, a2) = (a2, a1)
-                            match = (a1, a2, torsion[0], torsion[t4[1]], tordef)
+                            e4 = data.atoms[a4].element
+                            # the following for AMBER only - TODO: decide how to pass this in ffxml's
+                            if isAmber:
+                                if not hasWildcard:
+                                    if t2[0] == t4[0] and (r1 > r4 or (r1 == r4 and ta1 > ta4)):
+                                        (a1, a4) = (a4, a1)
+                                    if t3[0] == t4[0] and (r2 > r4 or (r2 == r4 and ta2 > ta4)):
+                                        (a2, a4) = (a4, a2)
+                                    if t2[0] == t3[0] and (r1 > r2 or (r1 == r2 and ta1 > ta2)):
+                                        (a1, a2) = (a2, a1)
```

Comment:
> I could perhaps change the translation code to:

Or alternatively

```Python
a2 = torsion[t2[1]]
a3 = torsion[t3[1]]
a4 = torsion[t4[1]]
```

The thing that makes it confusing is mixing two different numbering schemes.  Sticking to one or the other would be clearer.  It doesn't matter much which one, as long as it's consistent.  But if you write something like

```Python
(t1, t2) = (t2, t3)
```

that's clearly not being consistent.  It's repurposing the same set of variables that were initially named based on one numbering scheme, then reinterpreting the names as a different numbering scheme.